src/libopensc/card-epass2003.c

```diff
@@ -549,49 +693,83 @@ verify_init_key(struct sc_card *card, unsigned char *ran_key, unsigned char key_
 	unsigned char mac[256] = { 0 };
 	unsigned long i;
 	unsigned char tmp_sm;
+	unsigned char isFips;
 	epass2003_exdata *exdata = NULL;
 
 	if (!card->drv_data) 
 		return SC_ERROR_INVALID_ARGUMENTS;
 	exdata = (epass2003_exdata *)card->drv_data;
+	isFips = exdata->bFipsCertification;
 
 	LOG_FUNC_CALLED(card->ctx);
-
-	memcpy(data, ran_key, 8);
-	memcpy(&data[8], g_random, 8);
-	data[16] = 0x80;
-	memset(&data[17], 0x00, blocksize - 1);
-	memset(iv, 0, 16);
+	
+	if(isFips)
+	{
+		memset(data,0x00,15);
+        	memset(&data[11], 0x01, 1);
+        	memset(&data[14], 0x40, 1);
+        	memset(&data[15], 0x01, 1);
+        	memcpy(&data[16], g_random, 8);
+        	memcpy(&data[24], ran_key, 8);
+	}else{
+		memcpy(data, ran_key, 8);
+		memcpy(&data[8], g_random, 8);
+		data[16] = 0x80;
+		memset(&data[17], 0x00, blocksize - 1);
+		memset(iv, 0, 16);
+	}
 
 	/* calculate host cryptogram */
 	if (KEY_TYPE_AES == key_type) {
-		aes128_encrypt_cbc(exdata->sk_enc, 16, iv, data, 16 + blocksize,
-				   cryptogram);
+		if(isFips){
+			r = aes128_encrypt_cmac(exdata->sk_enc, 128, data, 32, cryptogram);
+		}
+		else{
+			r = aes128_encrypt_cbc(exdata->sk_enc, 16, iv, data, 16 + blocksize,cryptogram);
+		}
 	} else {
-		des3_encrypt_cbc(exdata->sk_enc, 16, iv, data, 16 + blocksize,
-				 cryptogram);
+		r = des3_encrypt_cbc(exdata->sk_enc, 16, iv, data, 16 + blocksize,cryptogram);
 	}
 
+	LOG_TEST_RET(card->ctx, r, "calculate host cryptogram  failed");
+
 	memset(data, 0, sizeof(data));
-	memcpy(data, "\x84\x82\x03\x00\x10", 5);
-	memcpy(&data[5], &cryptogram[16], 8);
-	memcpy(&data[13], "\x80\x00\x00", 3);
+	if(isFips){
+		memcpy(data, "\x84\x82\x03\x00\x10", 5);
+        	memcpy(&data[5], &cryptogram[0], 8);
+	}else{
+		memcpy(data, "\x84\x82\x03\x00\x10", 5);
```

Comment:
this line is same in both `if` and `else` branches. Can this be moved before the condition to avoid code duplication?

Reply:
not fixed

Reply:
sorry i have a mistake .repaired it .